tket2-hseries/src/extension/qsystem/barrier/qtype_analyzer.rs

```diff
@@ -0,0 +1,305 @@
+use hugr::types::{CustomType, SumType, Type, TypeArg};
+use std::collections::HashMap;
+
+/// If a type is an option of qubit.
+pub(crate) fn is_opt_qb(ty: &Type) -> bool {
+    if let Some(sum) = ty.as_sum() {
+        if let Some(inner) = sum.as_unary_option() {
+            return inner == &qb_t();
+        }
+    }
+    false
+}
+
+/// If a custom type is an array, return size and element type.
+pub(crate) fn array_args(ext: &CustomType) -> Option<(u64, &Type)> {
+    array::array_type_def()
+        .check_custom(ext)
+        .ok()
+        .and_then(|_| match ext.args() {
+            [TypeArg::BoundedNat { n }, TypeArg::Type { ty: elem_ty }] => Some((*n, elem_ty)),
+            _ => None,
+        })
+}
+
+/// Analyzes quantum types to determine how they should be unpacked
+/// for barrier insertion.
+pub(crate) struct QTypeAnalyzer {
+    /// Cache of unpacked types.
+    qubit_ports: HashMap<Type, Option<Vec<Type>>>,
+}
+
+impl QTypeAnalyzer {
+    /// Create a new instance of the [QTypeAnalyzer].
+    pub fn new() -> Self {
+        Self {
+            qubit_ports: HashMap::new(),
+        }
+    }
+
+    /// Compute the row produced when a type is unpacked.
+    /// Uses memoization to avoid recomputing the same type.
+    /// `None` if the type does not contain qubits.
+    pub fn unpack_type(&mut self, ty: &Type) -> Option<Vec<Type>> {
+        match self.qubit_ports.get(ty) {
+            Some(unpacked) => unpacked.clone(),
+            None => {
+                let unpacked = self._new_unpack_type(ty);
+                self.qubit_ports.insert(ty.clone(), unpacked.clone());
+                unpacked
+            }
+        }
+    }
+
+    /// Compute the row produced when a type is unpacked for the first time.
+    fn _new_unpack_type(&mut self, ty: &Type) -> Option<Vec<Type>> {
+        if ty == &qb_t() {
+            return Some(vec![qb_t()]);
+        }
+
+        if let Some(row) = ty.as_sum().and_then(SumType::as_tuple) {
+            let mut any_qb = false;
+            let unpacked_row = row
+                .iter()
+                .flat_map(|t| {
+                    let t = &t.clone().try_into_type().expect("unexpected row variable.");
+                    match self.unpack_type(t) {
+                        Some(inner) => {
+                            any_qb = true;
+                            inner
+                        }
+                        None => vec![t.clone()],
+                    }
+                })
+                .collect::<Vec<_>>();
+            any_qb.then_some(unpacked_row)
+
+            // other sums containing qubits are ignored.
+        } else if let Some((size, elem_ty)) = ty.as_extension().and_then(array_args) {
+            // Special case for Option[Qubit] since it is used in guppy qubit arrays.
+            // Fragile - would be better with dedicated guppy array type.
+            // Not sure how this can be improved without runtime barrier being able to
```

Comment:
Maybe worth making an issue for this?

Reply:
https://github.com/CQCL/tket2/issues/871